reader.go

```diff
@@ -284,39 +290,65 @@ func (r *reader) run(ctx context.Context, offset int64) {
 		// to the connection we know we'll want to restart from this offset.
 		offset = start
 
+		errcount := 0
 	readLoop:
 		for {
+			if !sleep(ctx, backoff(errcount, 100*time.Millisecond, 10*time.Second)) {
+				conn.Close()
+				return
+			}
+
 			switch offset, err = r.read(ctx, offset, conn); err {
 			case nil:
+				errcount = 0
+				continue
 			case RequestTimedOut:
-			case context.Canceled: // another reader has taken over
+				// Timeout on the kafka side, this can be safely retried.
+			case OffsetOutOfRange:
+				// We may be reading past the last offset, will retry later.
+			case context.Canceled:
+				// Another reader has taken over, we can safely quit.
 				conn.Close()
 				return
 			default:
-				conn.Close()
-				break readLoop
+				if _, ok := err.(Error); ok {
+					r.sendError(ctx, err)
+				} else {
+					conn.Close()
+					break readLoop
+				}
 			}
+
+			errcount++
 		}
 	}
 }
 
 func (r *reader) initialize(ctx context.Context, offset int64) (conn *Conn, start int64, err error) {
-	var whence int
-
-	if offset < 0 {
-		offset = 0
-	} else {
-		whence = 1
-	}
-
 	for _, broker := range r.brokers {
+		var first int64
+
 		if conn, err = r.dialer.DialLeader(ctx, "tcp", broker, r.topic, r.partition); err != nil {
 			continue
 		}
 
-		if start, err = conn.Seek(offset, whence); err != nil {
+		// This deadline controls how long the offset negotiation may take.
+		conn.SetDeadline(time.Now().Add(10 * time.Second))
+
+		if first, err = conn.ReadFirstOffset(); err != nil {
 			conn.Close()
-			continue
+			break
+		}
+
+		// In case the reader was configured with an offset before the first
+		// offset, skipping directly to the first offset.
+		if offset < first {
```

Comment:
Yes, you lost 100 messages that don't exist anymore anyway. That was my comment above, we should be notified/alerted but I don't think this should prevent reading (and we should be more explicit than just offset-out-of-range).

I'll think of something!